### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by Mingling94**
Lots of stuff, mostly just more experimentation

### CODE REVIEW

File: app/index.html [hunk 1]
```diff
@@ -43,13 +34,13 @@
...
+			<div class='pure-u-1-1 sidebar-fill'>
+			</div>
 		</div>
 
 		<!-- Mainbar -->
-		<div class='pure-u mainbar'>
-			<div id='mainbar-content'>
-				<div id='view'></div>
-			</div>
+		<div class='pure-u mainbar' style='height:100%' id='mainbar'>
```
Comment (DavidVorick) [inline app/index.html]: can you move the 'style' to the css file?

also, 'mainbar' should probably be renamed to 'webview' or 'webviewer' (if there's a keyword conflict or something)

File: app/js/UI.js [hunk 2]
```diff
@@ -2,17 +2,24 @@
...
+// Global variables and require statements available to all renderer processes
 'use strict';
 var fs = require('fs');
 var path = require('path');
+var webFrame = require('web-frame');
 
 // Defines the functions and private vars of UI.js
 var UI = (function() {
 
+	// Used to make the app more readable on hidpi screens. web-frame has
```
Comment (DavidVorick) [inline app/js/UI.js]: high dpi

also looks like the comment is unfinished

File: app/js/UI.js [hunk 3]
```diff
@@ -2,17 +2,24 @@
...
+var webFrame = require('web-frame');
 
 // Defines the functions and private vars of UI.js
 var UI = (function() {
 
+	// Used to make the app more readable on hidpi screens. web-frame has 
+	// TODO: Automate this based on screen size
+	var atomScreen = require('screen');
+	var screenSize = atomScreen.getPrimaryDisplay().workAreaSize;
+	webFrame.setZoomFactor(2);
```
Comment (DavidVorick) [inline app/js/UI.js]: ?

This doesn't interfere with low-pixel-count screens?

File: app/js/plugins.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
+// Global variables and require statements
+'use strict';
+
+// plugins manages all plugin logic on a more back-end level for the UI
+UI._plugins = (function(){
+	// Directory of plugins
+	var pluginsDir = path.join(__dirname, 'plugins');
+    // DOM element shortcuts
```
Comment (DavidVorick) [inline app/js/plugins.js]: comment is not tabbed properly

File: app/js/plugins.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	var head, sideBar, mainBar;
+	// Default plugin
+	var home;
+
+	// init performs startup logic related to plugins
+	function init(){
+
+		// Initialize variables upon document completion
+		head = document.getElementsByTagName('head')[0];
+        sideBar = document.getElementById('sidebar');
```
Comment (DavidVorick) [inline app/js/plugins.js]: also not tabbed well

File: app/js/plugins.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+				addButton(plugin);
+			});
+		});
+	}
+	
+    // initHome detects Overview or otherwise the first plugin, alphabetically,
+    // and loads it
+    function initHome(pluginNames) {
+
+        // Detect if Overview plugin is installed and set it to be the top
```
Comment (DavidVorick) [inline app/js/plugins.js]: no newline after function header

Comment (DavidVorick) [inline app/js/plugins.js]: (same for the other places)

File: app/js/plugins.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        // button and autoloaded view if so
+		var ovIndex = pluginNames.indexOf('Overview');
+		if (ovIndex !== -1 && pluginNames[0] !== 'Overview') {
+			pluginNames[ovIndex] = pluginNames[0];
+			pluginNames[0] = 'Overview';
+		}
+		home = pluginNames[0];
+	}
```
Comment (DavidVorick) [inline app/js/plugins.js]: we usually put helper functions above the function that calls them.

File: app/js/plugins.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+		mainBar = document.getElementById('mainbar');
+
+		// Initialize UI components that require a read of /app/plugins
+		fs.readdir(pluginsDir, function (err, pluginNames) {
+			if (err) {
+				throw err;
+			}
+
+			// Determine default plugin
+			initHome(pluginNames);
```
Comment (DavidVorick) [inline app/js/plugins.js]: doing the ordering in-place seems odd to me.

Comment (DavidVorick) [inline app/js/plugins.js]: I think it's fine for now, but typically I would probably use a function that returns a new array - it's just safer that way, and more clear that something is happening.

Would also be interesting to let plugins choose their own priority.

File: app/main.js [hunk 9]
```diff
@@ -27,8 +35,11 @@ function startMainWindow() {
 	// Load the index.html of the app.
 	mainWindow.loadUrl('file://' + __dirname + '/index.html');
 
+	// Focus selector on test-ui	
+	mainWindow.focus();
+
 	// DEVTOOL: Open the devtools.
-	// mainWindow.openDevTools();
+	mainWindow.openDevTools();
```
Comment (DavidVorick) [inline app/main.js]: whoops

File: .jshintrc [hunk 10]
```diff
@@ -117,9 +117,6 @@
 		//"iLoveDouglas"
```
Comment (DavidVorick) [inline .jshintrc]: ?

File: package.json [hunk 11]
```diff
@@ -6,19 +6,20 @@
...
-	"type": "git",
-	"url": "git://github.com/NebulousLabs/test-ui.git"
+    "type": "git",
+    "url": "git://github.com/NebulousLabs/test-ui.git"
   },
   "bugs": {
-	"url": "https://github.com/NebulousLabs/test-ui/issues"
+    "url": "https://github.com/NebulousLabs/test-ui/issues"
   }
 }
```
Comment (DavidVorick) [inline package.json:25]: is package.json supposed to be spaces and not tabs?

Comment (Mingling94) [inline package.json:25]: Huh, not sure how that happened. I'll edit it tomorrow.

File: app/main.js (hunk 9 above)
Comment (Mingling94) [inline app/main.js]: Hehehehe

File: app/js/UI.js (hunk 3 above)
Comment (Mingling94) [inline app/js/UI.js]: So this one's an interesting lack-of support within electron itself. I discussed it with people [on this closed thread in electron](https://github.com/atom/electron/issues/615#) and proposed up [this piece of code](https://github.com/atom/electron/issues/615#issuecomment-116366096) to deal with it. That zoom factor was for my own dev purposes. Let me know what you think

File: .jshintrc (hunk 10 above)
Comment (Mingling94) [inline .jshintrc]: XD, it's just leftover from the guy's .jshintrc that I copied this from. Douglas Crockford is the hardliner behind JSLint that JSHint is based on. I'll make it more professional.

File: app/index.html (hunk 1 above)
Comment (Mingling94) [inline app/index.html]: A keyword conflict with what though? Webview is itself a keyword used by electron (I think)

File: .jshintrc [hunk 12]
```diff
@@ -49,31 +49,31 @@
...
+	"asi"			: false,	// Tolerate Automatic Semicolon Insertion (no semicolons).
+	"boss"			: false,	// Tolerate assignments inside if, for & while. Usually conditions & loops are for comparison, not assignments.
+	"debug"			: false,	// Allow debugger statements e.g. browser breakpoints.
+	"eqnull"		: false,	// Tolerate use of `== null`.
+	"es5"			: false,	// Allow EcmaScript 5 syntax.
+	"esnext"		: false,	// Allow ES.next specific features such as `const` and `let`.
+	"evil"			: false,	// Tolerate use of `eval`.
+	"expr"			: false,	// Tolerate `ExpressionStatement` as Programs.
+	"funcscope"		: false,	// Tolerate declarations of variables inside of control structures while accessing them later from the outside.
+	"globalstrict"	: true,		// Allow global "use strict" (also enables 'strict').
```
Comment (DavidVorick) [inline .jshintrc:61]: missed one :P

File: .jshintrc [hunk 13]
```diff
@@ -81,27 +81,27 @@
...
-	"wsh"           : false,    // Enable globals available when your code is running as a script for the Windows Script Host.
+	"browser"		: true,		// Standard browser globals e.g. `window`, `document`.
+	"couch"			: false,	// Enable globals exposed by CouchDB.
+	"devel"			: false,	// Allow development statements e.g. `console.log();`.
+	"dojo"			: false,	// Enable globals exposed by Dojo Toolkit.
+	"jquery"		: true,		// Enable globals exposed by jQuery JavaScript library.
+	"mootools"		: false,	// Enable globals exposed by MooTools JavaScript framework.
+	"node"			: true,		// Enable globals available when code is running inside of the NodeJS runtime environment.
+	"nonstandard"	: false,	// Define non-standard but widely adopted globals such as escape and unescape.
+	"prototypejs"	: false,	// Enable globals exposed by Prototype JavaScript framework.
```
Comment (DavidVorick) [inline .jshintrc:92]: missed these two as well

File: app/js/UI.js [hunk 14]
```diff
@@ -1,23 +1,38 @@
...
-var fs = require('fs');
-var path = require('path');
-
-// Defines the functions and private vars of UI.js
 var UI = (function() {
+	// UI specific constants
+	// Constants used to calculate appropriate zoom
+	const screenSize = electronScreen.getPrimaryDisplay().workAreaSize;
+	const screenArea = screenSize.width * screenSize.height;
+	const screenUnit = 800 * 600;
```
Comment (DavidVorick) [inline app/js/UI.js]: 1024x768 is also an okay minimum.

It does notably break RPI's projectors, but it breaks little else. Most phones + netbooks are at least 1024x768

File: app/js/UI.js [hunk 15]
```diff
@@ -1,23 +1,38 @@
...
+
+	// configurable UI variables
+	// TODO: Allow these to be configurable by the user in some 'settings' page
+	var zoomScale = 2;
 
-	// Called at $(window).ready to initalize the view
+	// setLogScaleZoom makes the app more readable on high dpi screens. 
+	// TODO: Take better approach, resolution doesn't mean high dpi.
```
Comment (DavidVorick) [inline app/js/UI.js]: having the user set the zoom factor is okay/sufficient for other situations

Comment (DavidVorick) [inline app/js/UI.js]: at least for the short term

File: app/js/UI.js [hunk 16]
```diff
@@ -1,23 +1,38 @@
...
+	function setLogScaleZoom() {
+		var zoomFactor = Math.floor(util.log(zoomScale, screenArea / screenUnit));
+		webFrame.setZoomFactor(zoomFactor);
```
Comment (DavidVorick) [inline app/js/UI.js]: 1.5 is probably okay too

File: app/js/UI.js [hunk 17]
```diff
@@ -1,23 +1,33 @@
...
+
 // UI.js, the first renderer process, handles loading and transitioning between
 // buttons and views. Pretty much all user interaction response should go
 // through here.
+UI = (function() {
+	// UI specific constants
+	// Constants used to calculate appropriate zoom
+	const screenSize = electronScreen.getPrimaryDisplay().workAreaSize;
+	const screenArea = screenSize.width * screenSize.height;
+	const highRes = 3200 * 1745;
```
Comment (DavidVorick) [inline app/js/UI.js]: let's change this to > 2k, which is:

2048x1152

**Comment by Mingling94** [comment]
Can this be merged?

**Comment by DavidVorick** [comment]
Sorry I have been busy, but I will look at it shortly.

**Comment by DavidVorick** [comment]
Adam, Aaron, Luke, and soon me also have open pull requests

**Comment by DavidVorick** [comment]
If you are not busy right now, we could use some help with the code reviews

File: app/css/general.css [hunk 18]
```diff
@@ -78,17 +82,21 @@ html,body {
 	-webkit-user-select: text;
 }
 .sidebar-button:hover {
+	cursor: pointer;
 	background-color: #DDDDDD;
 }
 .sidebar-icon {
+	max-height: 80%;
+	width: auto;
```
Comment (DavidVorick) [inline app/css/general.css]: why set the max height? I'm not sure this is a portable way to do things

File: app/css/general.css [hunk 19]
```diff
@@ -78,17 +82,21 @@ html,body {
...
 }
 .sidebar-text {
+	padding: 4px;
 	padding-left: 8px;
 }
 
 .mainbar {
 	flex-grow: 1;
-
+	height: 100%;
```
Comment (DavidVorick) [inline app/css/general.css:97]: is there a reason this element is necessary?

File: app/js/UI.js [hunk 20]
```diff
@@ -1,23 +1,33 @@
+'use strict';
+
 // UI.js, the first renderer process, handles loading and transitioning between
 // buttons and views. Pretty much all user interaction response should go
 // through here.
+UI = (function() {
+	// UI specific constants
+	// Constants used to calculate appropriate zoom
+	const screenSize = electronScreen.getPrimaryDisplay().workAreaSize;
+	const screenArea = screenSize.width * screenSize.height;
```
Comment (DavidVorick) [inline app/js/UI.js:10]: I said app size, but screen size makes more sense. You don't want the scale changing as a person resizes their window. So this is fine.

**Comment by DavidVorick** [comment]
This looks good except for the poor compatibility with high resolution screens.

Adding a config folder with a scaling factor may be important sooner than otherwise anticipated. I think the only screen that really has a problem with the default zoom though is your yoga2

File: app/css/general.css (hunk 19 above)
Comment (Mingling94) [inline app/css/general.css:97]: webviews otherwise don't get a correct height (dunno exactly why this fixed it)

**Pull request merged**